test/atomistic.cpp

```diff
@@ -0,0 +1,269 @@
+#include <catch.hpp>
+
+#include <ovf.h>
+#include <fmt/format.h>
+
+
+#include <iostream>
+#include <cstring>
+
+void ovf_test_write(std::string filename, std::string meshtype, int ovf_extension_format)
+{
+    INFO( fmt::format("Testing 'ovf_test_write' with filename '{}', meshtype '{}', ovf_extension_format '{}'", filename, meshtype, ovf_extension_format) );
+
+    auto create_test_segment = [&]()
+    {
+        // segment header
+        auto segment = ovf_segment_create();
+        segment->title = const_cast<char *>("ovf test title - write");
+        segment->comment = const_cast<char *>("test write");
+        segment->valuedim   = 3;
+
+        segment->n_cells[0] = 2;
+        segment->n_cells[1] = 1;
+        segment->n_cells[2] = 1;
+
+        if(meshtype == "rectangular")
+            segment->n_cells[1] = 2;
+
+        // Fill in atomistic values
+        segment->bravaisa[0] = 1;
+        segment->bravaisa[1] = 0;
+        segment->bravaisa[2] = 0;
+
+        segment->bravaisb[0] = 0;
+        segment->bravaisb[1] = 1;
+        segment->bravaisb[2] = 0;
+
+        segment->bravaisc[0] = 0;
+        segment->bravaisc[1] = 0;
+        segment->bravaisc[2] = 1;
+
+        segment->pointcount = 4;
+
+        segment->ncellpoints = 2;
+        float basis[6] = {
+                            0,0.1,0.2,
+                            2,3,4
+                        };
+
+        segment->basis = (float *) malloc( segment->ncellpoints * 3 * sizeof(float) );
+        for( int ib=0; ib < segment->ncellpoints; ib++)
+        {
+            for(int i=0; i<3; i++)
+            {
+                segment->basis[ib*3 + i] = basis[ib*3 + i];
+            }
+        }
+
+        segment->N = 4;
+        segment->meshtype = strdup(meshtype.c_str());
+        return segment;
+    };
+
+    auto create_dummy_segment = [&]()
+    {
+        // Append second
+        auto segment = ovf_segment_create();
+        segment->title = const_cast<char *>("ovf test title - append");
+        segment->comment = const_cast<char *>("test append");
```

Comment:
`const_cast`ing here could yield undefined behaviour in other places, see https://en.cppreference.com/w/cpp/language/string_literal
It is used [here](https://github.com/spirit-code/spirit/blob/master/core/src/io/OVF_File.cpp#L18-L21) in Spirit as well...
I'm not sure what the right way to assign to these `char *`s would be in C++...

Seems like a wrapper like this might be the way to go:
```c++
template <size_t N>
char * ptr_from_literal(const char (&str)[N])
{
    char* ptr = new char[N];
    memcpy(ptr, str, N);
    return ptr;
}
```
(got it from https://www.py4u.net/discuss/101381 )

In fact, if we want to avoid a small memory leak in Spirit, we need to
```c++
if( segment->title )
    delete segment->title;
....
```
etc before each assignment to any of the `char *`s.